src/cargo/core/compiler/compilation.rs

```diff
@@ -82,6 +84,11 @@ pub struct Compilation<'cfg> {
     /// `RunCustomBuild` unit that generated these env vars.
     pub extra_env: HashMap<Metadata, Vec<(String, String)>>,
 
+    /// Environment variables generated artifact dependencies for consumption by future invocations of programs.
+    ///
+    /// The key is the artifact units metadata to uniquely identify the unit that created them.
+    pub artifact_env: HashMap<Metadata, HashSet<(String, PathBuf)>>,
```

Comment:
I don't know how this is going to be used, but `HashSet<(String, PathBuf)>` allows there to be two entries with the same `String` as long as they have different `PathBuf`. If these are env key value pairs, then that is going to be a hard bug to track down. `HashMap<String, PathBuf>`, may have semantics closer to what we want.

Reply:
That's a great catch - I also wouldn't see a reason to not make it a hashmap instead, which indeed is semantically what is intended:  a unique mapping between a single name and value.

The improvement is implemented here: 02fd23b4ee0de9920ec5e8c6023898cb991f763c